File: backend/signup.php
<?php
header("Access-Control-Allow-Origin: *");
header("Access-Control-Allow-Methods: POST");
header("Access-Control-Allow-Headers: Content-Type");

ini_set('display_errors', 1);
error_reporting(E_ALL);

// config.php: DB connection settings
include 'config.php';

// Get JSON data from React Front-End (POST request)
$data = json_decode(file_get_contents("php://input"));

$email = $data->email;
$password = $data->password;

// Validate input
if (empty($email) || empty($password)) {
    echo json_encode(["error" => "Email and Password are required."]);
    exit;
}

// Check if email already exists
$query = "SELECT * FROM users WHERE Email = ?";
$stmt = $conn->prepare($query);
$stmt->bind_param("s", $email);
$stmt->execute();
$result = $stmt->get_result();
if ($result->num_rows > 0) {
    echo json_encode(["error" => "Email already in use."]);
    exit;
}

// Hash the password
$passwordHash = password_hash($password, PASSWORD_DEFAULT);

// Insert user into the database
$query = "INSERT INTO users (Email, PasswordHash) VALUES (?, ?)";
$stmt = $conn->prepare($query);
$stmt->bind_param("ss", $email, $passwordHash);

if ($stmt->execute()) {
    echo json_encode(["message" => "Account created successfully!"]);
} else {
    echo json_encode(["error" => "An error occurred. Please try again."]);
}
?>

File: backend/signup.py
import os
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
import mysql.connector as sql
from argon2 import PasswordHasher

# Create a FastAPI router
router = APIRouter()

# Password hasher
ph = PasswordHasher()

# Define request model
class User(BaseModel):
    email: str
    password: str
    username: str

# Function to connect to the database
def get_db_connection():
    try:
        conn = sql.connect(
            host=os.getenv("DB_HOST", "dbhost.cs.man.ac.uk"),
            user=os.getenv("DB_USER", "y46354js"),
            password=os.getenv("DB_PASSWORD", "G53uPmjfOvBqLXrunGlcjdFRbSSxT9HtWk3P3oAEkTs"),
            database=os.getenv("DB_NAME", "2024_comp10120_cm5")
        )
        print("Connected to database successfully!")
        return conn
    except sql.Error as e:
        print(f"Database connection failed: {str(e)}")
        return None

# Signup route
@router.post("/signup")
async def signup(user: User):
    conn = get_db_connection()
    if not conn:
        raise HTTPException(status_code=500, detail="Database connection failed.")
    
    cursor = conn.cursor()

    # Check if email exists (Use `%s` for MySQL placeholders)
    cursor.execute("SELECT * FROM users WHERE Email = %s", (user.email,))
    existing_user = cursor.fetchone()

    if existing_user:
        cursor.close()
        conn.close()
        raise HTTPException(status_code=400, detail="Email already in use.")

    # Hash the password
    hashed_password = ph.hash(user.password)

    # Insert into database
    cursor.execute("INSERT INTO users (Email, PasswordHash, Username) VALUES (%s, %s, %s)", (user.email, hashed_password, user.username))
    conn.commit()
    
    cursor.close()
    conn.close()

    return {"message": "Account created successfully!"}